### ISSUE TITLE
Build failure on current Archlinux

### ISSUE THREAD

**Issue by zenhack** [sandstorm-io/sandstorm#3171]
I was trying to build from source just, and hit the error below. This is on current Archlinux. This bit looks relevant:

```
    In file included from /ekam-provider/canonical/sandstorm/web-session-bridge.c++:17:
    In file included from /ekam-provider/canonical/sandstorm/web-session-bridge.h:22:
    In file included from /ekam-provider/canonical/sandstorm/util.h:35:
    In file included from /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/set:61:
    /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_set.h:193:9: error: no matching member function for call to '_M_insert_range_unique'
            { _M_t._M_insert_range_unique(__first, __last); }
              ~~~~~^~~~~~~~~~~~~~~~~~~~~~
```

But I haven't been able to sort out what's going on. Here's the full set of errors:

```
   /ekam-provider/canonical/sandstorm/supervisor.c++:791:19: warning: 'readdir_r' is deprecated [-Wdeprecated-declarations]
          int error = readdir_r(dir, &entry, &eptr);
                      ^
    /usr/include/dirent.h:186:28: note: 'readdir_r' has been explicitly marked deprecated here
         __nonnull ((1, 2, 3)) __attribute_deprecated__;
                               ^
    /usr/include/sys/cdefs.h:251:51: note: expanded from macro '__attribute_deprecated__'
    # define __attribute_deprecated__ __attribute__ ((__deprecated__))
                                                      ^
    /ekam-provider/canonical/sandstorm/supervisor.c++:1903:29: warning: lambda capture 'this' is not used [-Wunused-lambda-capture]
        return req.send().then([this, context](auto args) mutable -> void {
                                ^~~~~
    In file included from /ekam-provider/canonical/sandstorm/supervisor.c++:17:
    In file included from /ekam-provider/canonical/sandstorm/supervisor.h:20:
    In file included from /ekam-provider/canonical/sandstorm/abstract-main.h:20:
    In file included from /ekam-provider/c++header/kj/main.h:28:
    In file included from /ekam-provider/c++header/kj/array.h:28:
    /ekam-provider/c++header/kj/memory.h:405:60: warning: delete called on non-final 'sandstorm::SupervisorMain::SupervisorImpl::LogWatcher' that has virtual functions but non-virtual destructor [-Wdelete-non-
      virtual-dtor]
      virtual void disposeImpl(void* pointer) const override { delete reinterpret_cast<T*>(pointer); }
                                                               ^
    /ekam-provider/c++header/kj/memory.h:411:51: note: in instantiation of member function 'kj::_::HeapDisposer<sandstorm::SupervisorMain::SupervisorImpl::LogWatcher>::disposeImpl' requested here
    const HeapDisposer<T> HeapDisposer<T>::instance = HeapDisposer<T>();
                                                      ^
    /ekam-provider/c++header/kj/memory.h:422:72: note: in instantiation of static data member 'kj::_::HeapDisposer<sandstorm::SupervisorMain::SupervisorImpl::LogWatcher>::instance' requested here
      return Own<T>(new T(kj::fwd<Params>(params)...), _::HeapDisposer<T>::instance);
                                                                           ^
    /ekam-provider/canonical/sandstorm/supervisor.c++:2099:24: note: in instantiation of function template specialization 'kj::heap<sandstorm::SupervisorMain::SupervisorImpl::LogWatcher, kj::UnixEventPort &, char
      const (&)[4], kj::AutoCloseFd, sandstorm::ByteStream::Client>' requested here
        auto watcher = kj::heap<LogWatcher>(eventPort, "log", kj::mv(logFile), params.getStream());
    ...(log truncated; use -l to increase log limit)...
✔ test: capnp/fuzz-test
✘ compile: sandstorm/web-session-bridge.c++
    /ekam-provider/canonical/sandstorm/web-session-bridge.c++:143:35: warning: lambda capture 'method' is not used [-Wunused-lambda-capture]
          auto doNonStreaming = [this,method,path,&headers,&requestBody,&response]() {
                                     ~^~~~~~
    /ekam-provider/canonical/sandstorm/web-session-bridge.c++:173:24: warning: lambda capture 'headers' is not used [-Wunused-lambda-capture]
              .then([this,&headers,&requestBody,&response,KJ_MVCAP(streamer)]
                         ~~^~~~~~~
    /ekam-provider/canonical/sandstorm/web-session-bridge.c++:177:11: warning: lambda capture 'this' is not used [-Wunused-lambda-capture]
          }, [this,KJ_MVCAP(doNonStreaming)](kj::Exception&& e) -> kj::Promise<void> {
              ^~~~~
    /ekam-provider/canonical/sandstorm/web-session-bridge.c++:192:35: warning: lambda capture 'method' is not used [-Wunused-lambda-capture]
          auto doNonStreaming = [this,method,path,&headers,&requestBody,&response]() {
                                     ~^~~~~~
    /ekam-provider/canonical/sandstorm/web-session-bridge.c++:222:24: warning: lambda capture 'headers' is not used [-Wunused-lambda-capture]
              .then([this,&headers,&requestBody,&response,KJ_MVCAP(streamer)]
                         ~~^~~~~~~
    /ekam-provider/canonical/sandstorm/web-session-bridge.c++:226:11: warning: lambda capture 'this' is not used [-Wunused-lambda-capture]
          }, [this,KJ_MVCAP(doNonStreaming)](kj::Exception&& e) -> kj::Promise<void> {
              ^~~~~
    In file included from /ekam-provider/canonical/sandstorm/web-session-bridge.c++:17:
    In file included from /ekam-provider/canonical/sandstorm/web-session-bridge.h:22:
    In file included from /ekam-provider/canonical/sandstorm/util.h:35:
    In file included from /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/set:61:
    /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_set.h:193:9: error: no matching member function for call to '_M_insert_range_unique'
            { _M_t._M_insert_range_unique(__first, __last); }
              ~~~~~^~~~~~~~~~~~~~~~~~~~~~
    /ekam-provider/canonical/sandstorm/util.h:244:30: note: in instantiation of function template specialization 'std::set<kj::StringPtr, std::less<kj::StringPtr>, std::allocator<kj::StringPtr> >::set<capnp::_::
      IndexingIterator<const capnp::List<capnp::Text, capnp::Kind::BLOB>::Reader, capnp::Text::Reader> >' requested here
      HeaderWhitelist(T&& list): patterns(list.begin(), list.end()) {}
                                 ^
    /ekam-provider/canonical/sandstorm/web-session-bridge.c++:98:7: note: in instantiation of function template specialization 'sandstorm::HeaderWhitelist::HeaderWhitelist<capnp::List<capnp::Text, capnp::Kind::
    ...(log truncated; use -l to increase log limit)...
✘ compile: sandstorm/sandstorm-http-bridge.c++
    /ekam-provider/canonical/sandstorm/sandstorm-http-bridge.c++:1520:10: warning: lambda capture 'this' is not used [-Wunused-lambda-capture]
            [this, KJ_MVCAP(httpRequest), KJ_MVCAP(clientStream), responseStream, context]
             ^~~~~
    /ekam-provider/canonical/sandstorm/sandstorm-http-bridge.c++:2468:37: warning: lambda capture 'this' is not used [-Wunused-lambda-capture]
        return address->connect().then([this, loggedSlowStartupMessage](auto x) -> void {
                                        ^~~~~
    In file included from /ekam-provider/canonical/sandstorm/sandstorm-http-bridge.c++:59:
    In file included from /ekam-provider/canonical/sandstorm/util.h:35:
    In file included from /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/set:61:
    /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_set.h:193:9: error: no matching member function for call to '_M_insert_range_unique'
            { _M_t._M_insert_range_unique(__first, __last); }
              ~~~~~^~~~~~~~~~~~~~~~~~~~~~
    /ekam-provider/canonical/sandstorm/util.h:244:30: note: in instantiation of function template specialization 'std::set<kj::StringPtr, std::less<kj::StringPtr>, std::allocator<kj::StringPtr> >::set<capnp::_::
      IndexingIterator<const capnp::List<capnp::Text, capnp::Kind::BLOB>::Reader, capnp::Text::Reader> >' requested here
      HeaderWhitelist(T&& list): patterns(list.begin(), list.end()) {}
                                 ^
    /ekam-provider/canonical/sandstorm/sandstorm-http-bridge.c++:154:23: note: in instantiation of function template specialization 'sandstorm::HeaderWhitelist::HeaderWhitelist<capnp::List<capnp::Text, capnp::
      Kind::BLOB>::Reader>' requested here
    const HeaderWhitelist REQUEST_HEADER_WHITELIST(*WebSession::Context::HEADER_WHITELIST);
                          ^
    /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_tree.h:1109:2: note: candidate template ignored: substitution failure [with _InputIterator = capnp::_::IndexingIterator<const 
      capnp::List<capnp::Text, capnp::Kind::BLOB>::Reader, capnp::Text::Reader>]: no type named 'value_type' in 'std::iterator_traits<capnp::_::IndexingIterator<const capnp::List<capnp::Text, capnp::Kind::BLOB>::
      Reader, capnp::Text::Reader> >'
            _M_insert_range_unique(_InputIterator __first, _InputIterator __last)
            ^
    /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_tree.h:1118:2: note: candidate template ignored: substitution failure [with _InputIterator = capnp::_::IndexingIterator<const 
      capnp::List<capnp::Text, capnp::Kind::BLOB>::Reader, capnp::Text::Reader>]: no type named 'value_type' in 'std::iterator_traits<capnp::_::IndexingIterator<const capnp::List<capnp::Text, capnp::Kind::BLOB>::
      Reader, capnp::Text::Reader> >'
            _M_insert_range_unique(_InputIterator __first, _InputIterator __last)
            ^
    ...(log truncated; use -l to increase log limit)...
✘ link: sandstorm/run-bundle.o
    /usr/bin/ld: tmp/kj/async-io-unix.o: in function `operator()':
    /ekam-provider/canonical/kj/async-io-unix.c++:989: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    /usr/bin/ld: tmp/sandstorm/gateway.o: in function `Tables':
    /ekam-provider/canonical/sandstorm/gateway.c++:92: undefined reference to `sandstorm::WebSessionBridge::Tables::Tables(kj::HttpHeaderTable::Builder&)'
    /usr/bin/ld: tmp/sandstorm/gateway.o: in function `sandstorm::GatewayService::request(kj::HttpMethod, kj::StringPtr, kj::HttpHeaders const&, kj::AsyncInputStream&, kj::HttpService::Response&)':
    /ekam-provider/canonical/sandstorm/gateway.c++:222: undefined reference to `sandstorm::WebSessionBridge::addStandardApiOptions(sandstorm::WebSessionBridge::Tables const&, kj::HttpHeaders const&, kj::
      HttpHeaders&)'
    /usr/bin/ld: tmp/sandstorm/gateway.o: in function `sandstorm::GatewayService::getStaticPublished(kj::StringPtr, kj::StringPtr, kj::HttpHeaders const&, kj::HttpService::Response&, unsigned int)':
    /ekam-provider/canonical/sandstorm/gateway.c++:759: undefined reference to `sandstorm::WebSessionBridge::makeHttpResponseStream(unsigned int, kj::StringPtr, kj::HttpHeaders&&, kj::HttpService::Response&)'
    /usr/bin/ld: tmp/sandstorm/gateway.o: in function `operator()':
    /ekam-provider/canonical/sandstorm/gateway.c++:270: undefined reference to `sandstorm::WebSessionBridge::addStandardApiOptions(sandstorm::WebSessionBridge::Tables const&, kj::HttpHeaders const&, kj::
      HttpHeaders&)'
    /usr/bin/ld: /ekam-provider/canonical/sandstorm/gateway.c++:489: undefined reference to `sandstorm::WebSessionBridge::restrictParentFrame(kj::StringPtr, kj::StringPtr)'
    /usr/bin/ld: tmp/sandstorm/gateway.o: in function `kj::Own<sandstorm::WebSessionBridge> kj::refcounted<sandstorm::WebSessionBridge, kj::Timer&, sandstorm::WebSession::Client, sandstorm::Handle::Client, 
      sandstorm::WebSessionBridge::Tables&, sandstorm::WebSessionBridge::Options&>(kj::Timer&, sandstorm::WebSession::Client&&, sandstorm::Handle::Client&&, sandstorm::WebSessionBridge::Tables&, sandstorm::
      WebSessionBridge::Options&)':
    /ekam-provider/c++header/kj/refcount.h:97: undefined reference to `sandstorm::WebSessionBridge::WebSessionBridge(kj::Timer&, sandstorm::WebSession::Client, kj::Maybe<sandstorm::Handle::Client>, sandstorm::
      WebSessionBridge::Tables const&, sandstorm::WebSessionBridge::Options)'
    /usr/bin/ld: tmp/sandstorm/gateway.o: in function `kj::Own<sandstorm::WebSessionBridge> kj::refcounted<sandstorm::WebSessionBridge, kj::Timer&, sandstorm::WebSession::Client, decltype(nullptr), sandstorm::
      WebSessionBridge::Tables&, sandstorm::WebSessionBridge::Options&>(kj::Timer&, sandstorm::WebSession::Client&&, decltype(nullptr)&&, sandstorm::WebSessionBridge::Tables&, sandstorm::WebSessionBridge::Options
      &)':
    /ekam-provider/c++header/kj/refcount.h:97: undefined reference to `sandstorm::WebSessionBridge::WebSessionBridge(kj::Timer&, sandstorm::WebSession::Client, kj::Maybe<sandstorm::Handle::Client>, sandstorm::
      WebSessionBridge::Tables const&, sandstorm::WebSessionBridge::Options)'
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
✘ install: sandstorm/sandstorm.ekam-manifest
    sandstorm/run-bundle: not found
vim 
```

**Comment by zenhack** [comment]
Figured it out. This is the relevant bit:

```
    /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_tree.h:1109:2: note: candidate template ignored: substitution failure [with _InputIterator = capnp::_::IndexingIterator<const 
      capnp::List<capnp::Text, capnp::Kind::BLOB>::Reader, capnp::Text::Reader>]: no type named 'value_type' in 'std::iterator_traits<capnp::_::IndexingIterator<const capnp::List<capnp::Text, capnp::Kind::BLOB>::
      Reader, capnp::Text::Reader> >'
            _M_insert_range_unique(_InputIterator __first, _InputIterator __last)
```

I got it to build successfully by adding `-DKJ_STD_COMPAT` to `CFLAGS`. Maybe there's somewhere in the source this ought to be defined, but I'm not sure where the right place would be.

**Comment by xet7** [comment]
@zenhack 

Maybe here?
https://github.com/sandstorm-io/sandstorm/blob/master/Makefile#L21

**Comment by zenhack** [comment]
Quoting Lauri Ojansivu (2019-12-05 00:10:44)

>    Maybe here?
>    https://github.com/sandstorm-io/sandstorm/blob/master/Makefile#L21

CFLAGS is supposed to be overridable from the CLI, so probably if it
were going to be hard-coded in the Makefile it should go in `CXXFLAGS2`,
which seems to be the place where non-conditional flags are put.

But I don't fully understand why this flag is necessary on my system and
(presumably) @kentonv's.

-Ian

**Comment by zenhack** [comment]
I'm running into two more build issues:

1. Node-fibers seems not to correctly work out the version of glibc in the resulting bundle
2. `node-capnp` ends up in the wrong place in the bundle.

node-fibers seems to be using [this package][1] to detect libc, and it's heuristics don't quite work for the bundle that gets built on my machine. I managed to hack around that by looking at the source to that package and monkeypatching the bundle so it would come up with the right answer (I just added a bin/getconf that returns the right answer). I'd like to come up with a better solution to this, obviously

I'm not sure why node-capnp isn't being found. the result is that this error ends up in sandstorm.log:


```
** Starting Sandstorm at: Sat Dec  7 21:37:39 2019
WARNING: Auto-updates are disabled by config.
** Starting back-end...
** Starting MongoDB...
note: noprealloc may hurt performance in many applications
about to fork child process, waiting until server is ready for connections.
forked process: 7
child process started successfully, parent exiting
** Back-end and Mongo started; now starting front-end...
** Starting Gateway...
Sun, 08 Dec 2019 02:37:43 GMT body-parser deprecated undefined extended: provide extended option at app/app.js:5506:39
Unhandled exception:  Error: Cannot find module "capnp"
    at Object.require (/programs/server/boot.js:303:24)
    at scheduled-job.js (server/scheduled-job.js:17:19)
    at fileEvaluate (packages/modules-runtime.js:336:7)
    at Module.require (packages/modules-runtime.js:238:14)
    at require (packages/modules-runtime.js:258:21)
    at /programs/server/app/app.js:18432:1
    at /programs/server/boot.js:419:36
    at Array.forEach (<anonymous>)
    at /programs/server/boot.js:228:19
    at /programs/server/boot.js:479:5
    at Function.run (/programs/server/profile.js:510:12)
    at /programs/server/boot.js:478:11
```
Interestingly, the bundle contains two `node_modules` directories, one at the root and one in `programs/server`. `node-capnp` seems to be located in the former, with everything else in the latter. This seems like the obvious culprit, but the release build (which works) is the same in this regard.

I tried building in a debian vm, which didn't have the problem with node-fibers but did have the problem with node-capnp.

Would appreciate suggestions on what to try (@kentonv?)

[1]: https://github.com/lovell/detect-libc

**Comment by kentonv** [comment]
This file is probably highly relevant to the problem finding capnp: https://github.com/sandstorm-io/sandstorm/blob/master/shell/packages/sandstorm-capnp/capnp.js

I don't understand why KJ_STD_COMPAT would be needed...

**Comment by zenhack** [comment]
Figured out the node-capnp thing. Turns out this was due to something specific to the cron branch that I'd merged in; fixed in this commit:

https://github.com/zenhack/sandstorm/commit/cdae2a294f86994e469c3430bb375c97c69efb5c

I feel a bit sheepish not having realized the last error didn't happen on master (though the other two do).

We don't seem to be doing Npm.require for capnp anywhere else in the source tree, except for the file you linked to. I don't really understand the magic that seems to be going on with imports, but I can at least successfully build the thing now.

**Comment by kentonv** [comment]
Yeah, we used to use `Npm.require("capnp")` all over the place until Meteor changed the path lookup semantics. See c4fb0d825c072220c2b26ccc0754a325e1283493 for where it was removed. I guess the cron PR predates that.

The `KJ_STD_COMPAT` thing is still strange to me. Seems like a bug in the version of libstdc++ on your machine -- apparently your implementation of `std::set::insert(begin,end)` requires iterators that have had type traits defined, but that seems like a glaring backwards-incompatibility...
